src/Raven.Server/Documents/Subscriptions/AbstractSubscriptionStorage.cs

```diff
@@ -0,0 +1,171 @@
+                if (subscriptionStateRaw == null)
+                {
+                    DeleteAndSetException(id, new SubscriptionDoesNotExistException($"The subscription {subscriptionName} had been deleted"));
+                    continue;
+                }
+
+                SubscriptionState subscriptionState = JsonDeserializationClient.SubscriptionState(subscriptionStateRaw);
+                if (subscriptionState.Disabled)
+                {
+                    DropSubscriptionConnections(id, new SubscriptionClosedException($"The subscription {subscriptionName} is disabled and cannot be used until enabled"));
+                    continue;
+                }
+
+
+                //make sure we only drop old connection and not new ones just arriving with the updated query
+                if (subscriptionConnectionsState != null && subscriptionState.Query != subscriptionConnectionsState.Query)
+                {
+                    DropSubscriptionConnections(id, new SubscriptionClosedException($"The subscription {subscriptionName} query has been modified, connection must be restarted", canReconnect: true));
+                    continue;
+                }
+
+                if (SubscriptionChangeVectorHasChanges(subscriptionConnectionsState, subscriptionState))
+                {
+                    DropSubscriptionConnections(id, new SubscriptionClosedException($"The subscription {subscriptionName} was modified, connection must be restarted", canReconnect: true));
+                    continue;
+                }
+
+                var whoseTaskIsIt = GetSubscriptionResponsibleNode(databaseRecord, subscriptionState);
+                if (whoseTaskIsIt != _serverStore.NodeTag)
+                {
+                    DropSubscriptionConnections(id,
+                        new SubscriptionDoesNotBelongToNodeException("Subscription operation was stopped, because it's now under a different server's responsibility"));
+                }
+            }
+        }
+    }
+
+    public TState GetSubscriptionConnectionsState<T>(TransactionOperationContext<T> context, string subscriptionName) where T : RavenTransaction
+    {
+        var subscriptionState = _serverStore.Cluster.Subscriptions.ReadSubscriptionStateByName(context, _databaseName, subscriptionName);
+
+        if (_subscriptions.TryGetValue(subscriptionState.SubscriptionId, out TState concurrentSubscription) == false)
+            return null;
+
+        return concurrentSubscription;
+    }
+
+    public bool TryEnterSubscriptionsSemaphore()
+    {
+        return _concurrentConnectionsSemiSemaphore.Wait(0);
+    }
+
+    public void ReleaseSubscriptionsSemaphore()
+    {
+        _concurrentConnectionsSemiSemaphore.Release();
+    }
+
+    public void LowMemory(LowMemorySeverity lowMemorySeverity)
+    {
+        foreach (var state in _subscriptions)
```

Comment:
if a shard get into a low mem, what will happened?

Reply:
in case there are inactive subscription connections it will dispose them